File: scripts/structures/channel.py
from dataclasses import dataclass
from typing import Union

from structures.output import Output
from structures.types import Byte, Dword


@dataclass
class Channel:
    envelope_index: Byte
    order_index: Byte
    oscillator_index: Byte
    pitch: Dword
    output_flag: Byte
    output: Output
    splitter: Dword

    def __init__(
        self,
        envelope_index: Union[int, Byte],
        order_index: Union[int, Byte],
        oscillator_index: Union[int, Byte],
        pitch: Union[int, Dword],
        output_flag: Union[int, Byte],
        output: Output,
        splitter: Union[int, Dword],
    ):
        self.envelope_index = Byte(envelope_index)
        self.order_index = Byte(order_index)
        self.oscillator_index = Byte(oscillator_index)
        self.pitch = Dword(pitch)
        self.output_flag = Byte(output_flag)
        self.output = output
        self.splitter = Dword(splitter)

File: tracker/structures/channel.cpp
#include "../general.hpp"
#include "../utils/file.hpp"
#include "channel.hpp"

void Channel::serialize(std::ofstream &file) const {
    write_data(file, this, sizeof(Channel));
}

Channel *Channel::deserialize(std::ifstream &file) {
    Channel *channel = resource_manager.allocate<Channel>();
    read_data(file, &channel->splitter, sizeof(channel->splitter));
    read_data(file, &channel->pitch, sizeof(channel->pitch));
    read_data(file, &channel->target, sizeof(channel->target));
    read_data(file, &channel->output_flag, sizeof(channel->output_flag));
    read_data(file, &channel->flag, sizeof(channel->flag));
    read_data(file, &channel->envelope_index, sizeof(channel->envelope_index));
    read_data(file, &channel->order_index, sizeof(channel->order_index));
    read_data(file, &channel->oscillator_index, sizeof(channel->oscillator_index));
    read_data(file, &channel->fraction, sizeof(channel->fraction));
    return channel;
}
